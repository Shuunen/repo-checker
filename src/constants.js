import { join } from 'path'

export const home = process.env.HOME
export const dataFileName = 'repo-checker.config.js'
export const repoCheckerPath = join(__dirname, '..')
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export const dataDefaults = {
  auto_merge: true,
  ban_sass: true,
  dev_deps_only: true,
  license: 'GPL-3.0',
  max_size_ko: 10,
  user_mail: 'romain.racamier@gmail.com',
  user_name: 'Romain Racamier-Lafon',
}
