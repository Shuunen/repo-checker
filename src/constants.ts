import { join } from 'path'

/* istanbul ignore next */
export const home = process.env.HOME ?? '' // when does HOME is not defined ?
export const dataFileName = 'repo-checker.config.js'
/* istanbul ignore next */
export const repoCheckerPath = process.env.pwd ?? process.cwd()
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export class ProjectData {
  auto_merge = true
  ban_sass = true
  dev_deps_only = true
  is_module = false
  license = 'GPL-3.0'
  max_size_ko = 1
  npm_package = false
  package_name = ''
  quiet = false
  repo_id = 'a-great-repo'
  use_typescript = false
  use_vue = false
  user_id = 'Shuunen'
  user_id_lowercase = 'shuunen'
  user_mail = 'romain.racamier@gmail.com'
  user_name = 'Romain Racamier-Lafon'
  web_published = false
  web_url = 'https://my-website.com'

  constructor (data: Partial<ProjectData> = {}) {
    Object.assign(this, data)
  }
}

export const dataDefaults = new ProjectData()
