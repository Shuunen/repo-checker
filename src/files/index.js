const CheckConfigs = require('./configs')
const CheckReadme = require('./readme')
const CheckPackage = require('./package')
const CheckLicense = require('./license')
const CheckTravis = require('./travis')
const CheckRenovate = require('./renovate')
const CheckEditorConfig = require('./editor-config')
const CheckNvmRc = require('./nvmrc')

module.exports = [CheckConfigs, CheckNvmRc, CheckReadme, CheckPackage, CheckLicense, CheckTravis, CheckRenovate, CheckEditorConfig]
