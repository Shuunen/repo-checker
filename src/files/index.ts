export * from './configs'
export * from './editor-config'
export * from './eslint'
export * from './gh-workflow'
export * from './license'
export * from './nvmrc'
export * from './package'
export * from './readme'
export * from './renovate'
export * from './repo-checker'
export * from './travis'
// dam lint
