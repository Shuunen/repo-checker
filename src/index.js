#!/usr/bin/env node

const arg = require('arg')
const path = require('path')

const check = require('./check')
const log = require('./logger')

function init () {
  const args = arg({ '--target': String, '--data': String, '--fix': Boolean }, { argv: process.argv.slice(2) })
  let target = args['--target']
  const doFix = args['--fix']
  if (!target) {
    log.info('\n', `no target specified via : --target=path/to/directory`)
    log.info(`targeting current directory...`, '\n')
    target = '.'
  }
  let data = {}
  if (args['--data']) {
    const p = path.join(__dirname, '..', args['--data'])
    log.info('loading data from', p)
    data = require(p)
  } else if (doFix) {
    log.warn('you should provide data to enhance fix')
  }
  log.start()
    .then(() => check(target, data, doFix))
    .catch(err => log.error(err))
    .then(() => log.end())
}

init()
