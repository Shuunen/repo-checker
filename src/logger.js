/* eslint-disable no-console */
const fs = require('fs')
const chalk = require('chalk')

const pkg = require('../package.json')

class Logger {
  get indent () {
    return '    '.repeat(this.indentLevel)
  }

  get date () {
    return new Date().toISOString().replace('T', ' ').split('.')[0]
  }

  constructor (filePath) {
    this.file = fs.createWriteStream(filePath, { flags: 'a' })
    this.indentLevel = 0
  }

  async _write (...args) {
    this.file.write(args.join(' ') + '\n')
  }

  async setIndentLevel (level) {
    this.indentLevel = level
  }

  debug (...args) {
    return this._write(this.indent, 'üîπ', ...args)
  }

  info (...args) {
    console.log(this.indent, '‚¨ú ', ...args)
    return this._write(this.indent, '‚¨ú', ...args)
  }

  error (...args) {
    console.error(chalk.redBright(this.indent, '‚ùå ', ...args))
    return this._write(this.indent, '‚ùå', ...args)
  }

  warn (...args) {
    console.log(chalk.yellowBright(this.indent, '‚ö†Ô∏è ', ...args))
    return this._write(this.indent, '‚ö†Ô∏è', ...args)
  }

  success (...args) {
    console.log(chalk.green(this.indent, '‚úîÔ∏è ', ...args))
    return this._write(this.indent, '‚úîÔ∏è', ...args)
  }

  test (ok, msg, justWarn) {
    if (ok) {
      this.success(msg)
    } else if (justWarn) {
      this.warn(msg)
    } else {
      this.error(msg)
    }
  }

  fix (...args) {
    console.log(chalk.blueBright(this.indent, '‚¨ú ', ...args))
    return this._write(this.indent, '‚¨ú', ...args)
  }

  line () {
    console.log('')
    return this._write('')
  }

  start () {
    this.line()
    this._write(`‚¨áÔ∏è--- Entry from ${this.date} ---‚¨áÔ∏è`)
    this.info(`${pkg.name} v${pkg.version} is starting`)
    return this.line()
  }

  end () {
    this.line()
    this.info(pkg.name, 'has finished')
    return this.line()
  }
}

const log = new Logger(pkg.config.logFile)

module.exports = log
