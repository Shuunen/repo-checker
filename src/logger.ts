import { bgBlue, black, Logger } from 'shuutils'

class ExtendedLogger extends Logger {
  public fix (...stuff: readonly string[]): void {
    const items = Array.from(stuff)
    items.push(bgBlue(black('[ fixed ]')))
    this.info(...items)
  }

  public unknownError (error: unknown): void {
    if (error instanceof Error) { this.error(error.message); return }
    if (typeof error === 'string') { this.error(error); return }
    this.error('Unknown error format, could not log it')
  }
}

export const log = new ExtendedLogger()
