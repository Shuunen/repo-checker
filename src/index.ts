#!/usr/bin/env node

import { Nb } from 'shuutils'
import { log } from './logger'
import { start } from './main'

// eslint-disable-next-line unicorn/prefer-top-level-await
start().catch((error: unknown) => {
  log.unknownError(error)
  process.exit(Nb.One)
})
