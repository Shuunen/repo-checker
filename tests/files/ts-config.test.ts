import { Nb } from 'shuutils'
import { test } from 'uvu'
import { equal } from 'uvu/assert'
import { ProjectData } from '../../src/constants'
import { TsConfigFile } from '../../src/files'
import { log } from '../../src/logger'
import { promiseVoid } from '../utils'

test('ts config file no check', async function () {
  log.disable()
  const instance = new TsConfigFile('', new ProjectData({ isUsingTypescript: false }))
  await instance.start()
  await instance.end()
  equal(instance.passed, [])
  equal(instance.failed, [])
})

test('ts config file fix', async function () {
  log.disable()
  const instance = new TsConfigFile('', new ProjectData({ isUsingTypescript: true }), true)
  const fileInitial = '{ "name": "John", "files": ["src/main.ts"] }'
  const fileFixed = JSON.stringify({
    name: 'John',
    files: ['src/main.ts'],
    include: ['src'],
    compilerOptions: {
      /* eslint-disable @typescript-eslint/naming-convention */
      allowUnreachableCode: false,
      allowUnusedLabels: false,
      checkJs: true,
      esModuleInterop: true,
      exactOptionalPropertyTypes: true,
      forceConsistentCasingInFileNames: true,
      noFallthroughCasesInSwitch: true,
      noImplicitAny: true,
      noImplicitOverride: true,
      noImplicitReturns: true,
      noPropertyAccessFromIndexSignature: false,
      noUncheckedIndexedAccess: true,
      noUnusedLocals: true,
      noUnusedParameters: true,
      skipLibCheck: true,
      strict: true,
      outDir: './dist',
      moduleResolution: 'Node',
      target: 'ES2020',
      lib: ['ESNext'],
      types: [],
      /* eslint-enable @typescript-eslint/naming-convention */
    },
  }, undefined, Nb.Two)
  instance.inspectFile = promiseVoid
  instance.fileContent = fileInitial
  instance.updateFile = promiseVoid
  equal(instance.fileContent, fileInitial)
  await instance.start()
  await instance.end()
  equal(instance.fileContent, fileFixed)
  equal(instance.passed, [
    'does-not-use-wildcard-in-files-section',
    'my-folder-is-not-needed-in-include-section-my-folder-is-enough',
  ], 'passed')
  equal(instance.failed, [], 'failed')
})

test('ts config malformed', async function () {
  log.disable()
  const instance = new TsConfigFile('', new ProjectData({ isUsingTypescript: true }))
  const fileInitial = '"name": "John" }'
  instance.inspectFile = promiseVoid
  instance.fileContent = fileInitial
  equal(instance.fileContent, fileInitial)
  await instance.start()
  await instance.end()
  equal(instance.passed, [])
  equal(instance.failed, [])
})

test.run()
