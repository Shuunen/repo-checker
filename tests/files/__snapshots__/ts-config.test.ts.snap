// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ts config file fix 1`] = `
"{
  \\"name\\": \\"John\\",
  \\"files\\": [
    \\"src/main.ts\\"
  ],
  \\"include\\": [
    \\"src\\"
  ],
  \\"compilerOptions\\": {
    \\"allowUnreachableCode\\": false,
    \\"allowUnusedLabels\\": false,
    \\"checkJs\\": true,
    \\"esModuleInterop\\": true,
    \\"exactOptionalPropertyTypes\\": true,
    \\"forceConsistentCasingInFileNames\\": true,
    \\"noFallthroughCasesInSwitch\\": true,
    \\"noImplicitAny\\": true,
    \\"noImplicitOverride\\": true,
    \\"noImplicitReturns\\": true,
    \\"noPropertyAccessFromIndexSignature\\": false,
    \\"noUncheckedIndexedAccess\\": true,
    \\"noUnusedLocals\\": true,
    \\"noUnusedParameters\\": true,
    \\"skipLibCheck\\": true,
    \\"strict\\": true,
    \\"outDir\\": \\"./dist\\",
    \\"moduleResolution\\": \\"Node\\",
    \\"target\\": \\"ES2020\\",
    \\"lib\\": [
      \\"ESNext\\"
    ]
  }
}"
`;

exports[`ts config file fix 2`] = `
{
  "canFix": true,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": false,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": true,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "{
  \\"name\\": \\"John\\",
  \\"files\\": [
    \\"src/main.ts\\"
  ],
  \\"include\\": [
    \\"src\\"
  ],
  \\"compilerOptions\\": {
    \\"allowUnreachableCode\\": false,
    \\"allowUnusedLabels\\": false,
    \\"checkJs\\": true,
    \\"esModuleInterop\\": true,
    \\"exactOptionalPropertyTypes\\": true,
    \\"forceConsistentCasingInFileNames\\": true,
    \\"noFallthroughCasesInSwitch\\": true,
    \\"noImplicitAny\\": true,
    \\"noImplicitOverride\\": true,
    \\"noImplicitReturns\\": true,
    \\"noPropertyAccessFromIndexSignature\\": false,
    \\"noUncheckedIndexedAccess\\": true,
    \\"noUnusedLocals\\": true,
    \\"noUnusedParameters\\": true,
    \\"skipLibCheck\\": true,
    \\"strict\\": true,
    \\"outDir\\": \\"./dist\\",
    \\"moduleResolution\\": \\"Node\\",
    \\"target\\": \\"ES2020\\",
    \\"lib\\": [
      \\"ESNext\\"
    ]
  }
}",
  "fileContentObject": {
    "compilerOptions": {
      "allowUnreachableCode": false,
      "allowUnusedLabels": false,
      "checkJs": true,
      "esModuleInterop": true,
      "exactOptionalPropertyTypes": true,
      "forceConsistentCasingInFileNames": true,
      "lib": [
        "ESNext",
      ],
      "moduleResolution": "Node",
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitOverride": true,
      "noImplicitReturns": true,
      "noPropertyAccessFromIndexSignature": false,
      "noUncheckedIndexedAccess": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "outDir": "./dist",
      "skipLibCheck": true,
      "strict": true,
      "target": "ES2020",
    },
    "files": [
      "src/main.ts",
    ],
    "include": [
      "src",
    ],
    "name": "John",
  },
  "fileName": "",
  "originalFileContent": "",
  "passed": [
    "does-not-use-wildcard-in-files-section",
    "my-folder-is-not-needed-in-include-section-my-folder-is-enough",
  ],
  "warnings": [
    "could-have-a-non-empty-types-compiler-option-ex-types-node-fs-promises",
  ],
}
`;

exports[`ts config file no check 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": false,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [],
}
`;

exports[`ts config malformed 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": false,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": true,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "\\"name\\": \\"John\\" }",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [],
}
`;
